src/runtime/contrib/cudnn/conv_forward.cc

```diff
@@ -120,137 +200,144 @@ TVM_REGISTER_GLOBAL("tvm.contrib.cudnn.conv2d.forward")
-  int x_dim2 = args[9];
-  int x_dim3 = args[10];
-  int w_dim0 = args[11];
-  int w_dim1 = args[12];
-  int w_dim2 = args[13];
-  int w_dim3 = args[14];
-  void *out_shape = args[15];
-  std::string data_dtype = args[16];
-  std::string conv_dtype = args[17];
+
   // Set Data Type
   entry_ptr->conv_entry.data_type = CuDNNDataType::DLTypeToCuDNNType(String2TVMType(conv_dtype));
   cudnnDataType_t data_type = CuDNNDataType::DLTypeToCuDNNType(String2TVMType(data_dtype));
   // Set Format
   entry_ptr->conv_entry.tensor_format = static_cast<cudnnTensorFormat_t>(format);
+
   // conv desc
-  CUDNN_CALL(cudnnSetConvolution2dDescriptor(entry_ptr->conv_entry.conv_desc,
-                                             pad_h,
-                                             pad_w,
-                                             stride_h,
-                                             stride_w,
-                                             dilation_h,
-                                             dilation_w,
+  CUDNN_CALL(cudnnSetConvolutionNdDescriptor(entry_ptr->conv_entry.conv_desc,
+                                             dims,
+                                             pad_v,
+                                             stride_v,
+                                             dilation_v,
                                              CUDNN_CROSS_CORRELATION,
                                              entry_ptr->conv_entry.data_type));
-  // input desc
-  CUDNN_CALL(cudnnSetTensor4dDescriptor(entry_ptr->conv_entry.input_desc,
-                                        entry_ptr->conv_entry.tensor_format,
-                                        data_type,
-                                        x_dim0,
-                                        x_dim1,
-                                        x_dim2,
-                                        x_dim3));
-  // filter desc
-  CUDNN_CALL(cudnnSetFilter4dDescriptor(entry_ptr->conv_entry.filter_desc,
-                                        data_type,
-                                        entry_ptr->conv_entry.tensor_format,
-                                        w_dim0,
-                                        w_dim1,
-                                        w_dim2,
-                                        w_dim3));
-
-  CUDNN_CALL(cudnnGetConvolution2dForwardOutputDim(entry_ptr->conv_entry.conv_desc,
-                                                   entry_ptr->conv_entry.input_desc,
-                                                   entry_ptr->conv_entry.filter_desc,
-                                                   static_cast<int*>(out_shape),
-                                                   static_cast<int*>(out_shape) + 1,
-                                                   static_cast<int*>(out_shape) + 2,
-                                                   static_cast<int*>(out_shape) + 3));
-});
 
+  if (dims == 2 && entry_ptr->conv_entry.tensor_format ==  CUDNN_TENSOR_NHWC) {
```

Comment:
why do we need this special case?

Reply:
without this special case, I always got below error:
```
E             [bt] (2) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(TVMFuncCall+0x90) [0x3fff7599dfa0]
E             [bt] (1) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(+0xeea644) [0x3fff75a5a644]
E             [bt] (0) /host/root/ligc/tmp/incubator-tvm/build/libtvm.so(tvm::contrib::OutputShape(int, int, int const*, int const*, int const*, int const*, int const*, void*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)+0x1114) [0x3fff75a59a94]
E             File "/host/root/ligc/tmp/incubator-tvm/src/runtime/contrib/cudnn/conv_forward.cc", line 261
E           cuDNN: Check failed: e == CUDNN_STATUS_SUCCESS (3 vs. 0) : CUDNN_STATUS_BAD_PARAM
```
I have double checked the shape for input and weight, they should follow the SDK spec, would you please also take a look to see if any wrong in the code:
```
format

    Input.Type of the filter layout format. If this input is set to CUDNN_TENSOR_NCHW, which is one of the enumerant values allowed by cudnnTensorFormat_t descriptor, then the layout of the filter is as follows:

        For N=4, a 4D filter descriptor, the filter layout is in the form of KCRS:
            K represents the number of output feature maps
            C is the number of input feature maps
            R is the number of rows per filter
            S is the number of columns per filter
        For N=3, a 3D filter descriptor, the number S (number of columns per filter) is omitted.
        For N=5 and greater, the layout of the higher dimensions immediately follow RS.

    On the other hand, if this input is set to CUDNN_TENSOR_NHWC, then the layout of the filter is as follows:

        For N=4, a 4D filter descriptor, the filter layout is in the form of KRSC.
        For N=3, a 3D filter descriptor, the number S (number of columns per filter) is omitted and the layout of C immediately follows R.
        For N=5 and greater, the layout of the higher dimensions are inserted between S and C. For more information, see cudnnTensorFormat_t.
```